commit 9688071679a7084c82716aa1275b8965590ea759
Author: Christoph M. Becker <cmbecker69@gmx.de>
Date:   Tue Apr 6 12:44:42 2021 +0200

    Fix #80880: SSL_read on shutdown, ftp/proc_open
    
    When `SSL_read()` after `SSL_shutdown()` fails with `SSL_ERROR_SYSCALL`,
    we should not warn about this, because it is likely caused by the peer
    having closed the connection without having sent a close_notify
    shutdown alert.
    
    Signed-off-by: Christoph M. Becker <cmbecker69@gmx.de>
    
    Closes GH-6803.


diff --git a/ext/ftp/ftp.c b/ext/ftp/ftp.c
index 5bb359b353..f2da6e5153 100644
--- a/ext/ftp/ftp.c
+++ b/ext/ftp/ftp.c
@@ -1931,6 +1931,12 @@ static void ftp_ssl_shutdown(ftpbuf_t *ftp, php_socket_t fd, SSL *ssl_handle) {
 					/* SSL wants a write. Really odd. Let's bail out. */
 					done = 1;
 					break;
+				case SSL_ERROR_SYSCALL:
+					/* most likely the peer closed the connection without
+					   sending a close_notify shutdown alert;
+					   bail out to avoid raising a spurious warning */
+					done = 1;
+					break;
 				default:
 					if ((sslerror = ERR_get_error())) {
 						ERR_error_string_n(sslerror, buf, sizeof(buf));
